File: ctci/ch2/1/removedups.cpp
#include <unordered_set>
#include <iostream>

template<typename T> 
struct Node {
  T data;
  Node* next;
};

void removeNode(Node<int>* current, Node<int>* prev) {
  prev->next = current->next;
  delete current;
}

void removeDuplicates(Node<int>* head) {
  std::unordered_set<int> s;
  Node<int>* prev = NULL;
  Node<int>* current = head;
  while (current != NULL) {
    if (s.find(current->data) != s.end()) {
      removeNode(current, prev);
      current = prev->next;
    }
    else {
      s.insert(current->data);
      prev = current;
      current = current->next;
    }
  }
}

int main() {
  Node<int>* head = new Node<int>();
  head->data = 10;
  head->next = new Node<int>();
  head->next->data = -5;
  head->next->next = new Node<int>();
  head->next->next->data = -5;
  head->next->next->next = new Node<int>();
  head->next->next->next->data = 4;
  removeDuplicates(head);
  Node<int>* temp = head;
  while (temp != NULL) {
    std::cout << temp->data << "\n";
    temp = temp->next;
  }
}

File: ctci/ch2/1/RemoveDups.java
class Node {
    Node next = null;
    int data;

    public Node(int d) {
	data = d;
    }

    void appendToTail(int d) {
	Node end = new Node(d);
	Node n = this;
	while (n.next != null) {
	    n = n.next;
	}
	n.next = end;
    }

    static Node deleteNode(Node head, int d) {
	Node n = head;

	if (n.data == d) {
	    return head.next;
	}

	while (n.next != null) {
	    if (n.next.data == d) {
		n.next = n.next.next;
		return head;
	    }
	    n = n.next;	    
	}
	return head;
    }

    static void printNodes(Node head) {
	Node n = head;

	while (n != null) {
	    System.out.println(n.data);
	    n = n.next;
	}
    }
}

class RemoveDups {

    List<int> elementList = new ArrayList<>();
    
    public static void main(String[] args) {
	Node head = new Node(5);
	head.appendToTail(5);
	head.appendToTail(6);
	head.appendToTail(7);
	head.appendToTail(7);
	head.appendToTail(8);
	Node.printNodes(head);
    }
}
